### PR TITLE
finish creation of Template DB

### LINKED ISSUE #192: use Template Database as an end-to-end test
## motivations

we'd like to be able to run an end-to-end test of our build pipeline

this probably should
- include the major stages of all builds: ~Extract~, Load, Transform, Export
  - Extract is currently done by data-library, our builds expect source data to have been extracted/archived
- cover some the most common approaches seen in data products
- have realistic and fun/interesting outputs 📊

this probably shouldn't
- be expected to test every `dcpy` function or every single approach seen in our data products (e.g. a SQL procedure to apply corrections)
- take long to build
- force significant changes to the rest of the codebase (yet)

## thoughts

this is meant to test our data platform, not a particular data product

seems like we're most interested in covering the Load and Transform stages. Extract is performed by [`data-library`](https://github.com/NYCPlanning/db-data-library/)

started [a TDB wiki page](https://github.com/NYCPlanning/data-engineering/wiki/Product:-TDB) to start ideas for this fictional data product and perhaps be a model for what we'd like data product wikis to be

future stuff:
- be our most "up-to-date" product by being a codebase where we can implement our latest favored build approaches

## progress
- [x] part 1: add test structure and minimal Template DB build
- [ ] part 2: expand and test Template DB build

### PR DESCRIPTION
**Pull request by damonmcc**
resolves #192 

## goals
- add transform and export stages to TDB so that it can be a minimal model of our product build approaches
- create a mock data product that isn't very complex and is potentially fun/interesting to DE
- create a foundation for refining/expanding TDB later

## changes
- add a [Product design](https://github.com/NYCPlanning/data-engineering/wiki/Product:-TDB#product-design) section to the TDB wiki to brainstorm the design of this mock product
- change the TDB end-to-end tests to reflect the product design
- use new `postgres.PostgresClient` and the persistent DB server in the build
  - had to create a `db-template` DB in `edm-date`. named the DB to align with our product names in the `edm-publishing` S3 bucket
- shorten the list of source datasets
- add the transformation stage
- add the export stage
- lint all TDB sql files in PR tests
- update the [TDB wiki](https://github.com/NYCPlanning/data-engineering/wiki/Product:-TDB) to be more about product details and less about internal/dev
- update the [TDB readme](https://github.com/NYCPlanning/data-engineering/blob/template-db-build/products/template/README.md) to be more about internal/dev details and less about product details

### CODE REVIEW

File: products/template/tests/test_build.py
Comment (damonmcc) [inline products/template/tests/test_build.py:1]: these are the end-to-end tests. they illustrate the tables and files which are expected to exist after a build

File: products/template/sql/templatedb.sql
Comment (damonmcc) [inline products/template/sql/templatedb.sql:1]: this sql scripts generates the final primary table

File: products/template/seeds/boroughs.csv
Comment (damonmcc) [inline products/template/seeds/boroughs.csv:1]: this is a csv for Borough-related mappings. it's inspired by the concept of [seed files](https://docs.getdbt.com/docs/build/seeds#overview)

**Review by alexrichey — APPROVED** [review]
This is great. Looking forward to making use of this to test changes. I do wonder if we want to delete the existing s3 artifacts at the beginning of the test, to avoid false positives for successive test runs on a branch. (unless you're doing that already, and I missed it) But it's not urgent.

**Comment by damonmcc** [comment]
@alexrichey good call! I got false positives when I reset and pushed one commit at a time expecting more tests to fail

I'll note this and other enhancements in a follow-up issue (duplicative landmarks records, determining the boroughs of green spaces, etc.)

**Review by fvankrieken — APPROVED** [review]

**Pull request merged**